[C10D] Rewrite TCPStore client send path to minimize amount of syscalls.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100768
* __->__ #100742

Accumulate data in a local buffer prior to sending it. This reduces
the number of syscalls and network packets.

We flush every 1440 bytes to cap the amount of temporaty memory.

kumpera committed:
[C10D] Rewrite TCPStore client send path to minimize amount of syscalls.

Accumulate data in a local buffer prior to sending it. This reduces
the number of syscalls and network packets.

We flush every 1440 bytes to cap the amount of temporaty memory.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Rewrite TCPStore client send path to minimize amount of syscalls."


Accumulate data in a local buffer prior to sending it. This reduces
the number of syscalls and network packets.

We flush every 1440 bytes to cap the amount of temporaty memory.

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -1036,6 +997,52 @@ void TCPCallbackClient::setCallback(
+ public:
+  SendBuffer(detail::TCPClient& client, detail::QueryType cmd)
+      : client(client) {
+    buffer.reserve(32); // enough for most commands
```
Can we make this a const or adjustable rather than hard code it?

fduwjj reviewed (approved):
This is amazing! Thanks!

fduwjj reviewed (approved):
This is amazing! Thanks!

fduwjj reviewed (approved):
This is amazing! Thanks!

kumpera committed:
Update on "[C10D] Rewrite TCPStore client send path to minimize amount of syscalls."


Accumulate data in a local buffer prior to sending it. This reduces
the number of syscalls and network packets.

We flush every 1440 bytes to cap the amount of temporaty memory.

[ghstack-poisoned]

kumpera committed:
Update on "[C10D] Rewrite TCPStore client send path to minimize amount of syscalls."


Accumulate data in a local buffer prior to sending it. This reduces
the number of syscalls and network packets.

We flush every 1440 bytes to cap the amount of temporaty memory.

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge
